File: backend/src/FormaGeo.Application/SiteImports/GeoJsonImportParser.cs
using System.Text.Json;
using FormaGeo.Application.Sites.Contracts;
using FormaGeo.Application.Sites.Mapping;
using NetTopologySuite.Geometries;

namespace FormaGeo.Application.SiteImports;

public sealed record ParsedGeoJsonImport(
    string DetectedCoordinateSystem,
    IReadOnlyList<ParsedGeoJsonFeature> Features);

public sealed record ParsedGeoJsonFeature(
    int Index,
    string Name,
    IReadOnlyList<Polygon> Polygons,
    string? Error,
    IReadOnlyList<string> Warnings);

public static class GeoJsonImportParser
{
    private const double EarthRadiusMetres = 6_378_137;

    public static ParsedGeoJsonImport Parse(
        string content,
        string fallbackName)
    {
        JsonDocument document;

        try
        {
            document = JsonDocument.Parse(content);
        }
        catch (JsonException exception)
        {
            throw Invalid(
                "content",
                "invalid_geojson",
                $"The file is not valid JSON: {exception.Message}");
        }

        using (document)
        {
            var root = document.RootElement;

            if (root.ValueKind != JsonValueKind.Object)
            {
                throw Invalid(
                    "content",
                    "invalid_geojson",
                    "The GeoJSON root must be an object.");
            }

            var sourceSrid = DetectSrid(root);
            var coordinateSystem = sourceSrid == 3857
                ? "Web Mercator (EPSG:3857), transformed to WGS 84"
                : "WGS 84 (EPSG:4326)";
            var features = ReadFeatures(
                root,
                fallbackName,
                sourceSrid);

            return new ParsedGeoJsonImport(
                coordinateSystem,
                features);
        }
    }

    private static IReadOnlyList<ParsedGeoJsonFeature> ReadFeatures(
        JsonElement root,
        string fallbackName,
        int sourceSrid)
    {
        var type = ReadRequiredString(root, "type");

        if (string.Equals(
            type,
            "FeatureCollection",
            StringComparison.OrdinalIgnoreCase))
        {
            if (!root.TryGetProperty("features", out var featuresElement) ||
                featuresElement.ValueKind != JsonValueKind.Array)
            {
                throw Invalid(
                    "content",
                    "invalid_feature_collection",
                    "A FeatureCollection must contain a features array.");
            }

            return featuresElement
                .EnumerateArray()
                .Select((feature, index) =>
                    ReadFeature(
                        feature,
                        index,
                        fallbackName,
                        sourceSrid))
                .ToArray();
        }

        if (string.Equals(
            type,
            "Feature",
            StringComparison.OrdinalIgnoreCase))
        {
            return
            [
                ReadFeature(
                    root,
                    0,
                    fallbackName,
                    sourceSrid)
            ];
        }

        return
        [
            ReadGeometryFeature(
                root,
                0,
                fallbackName,
                sourceSrid)
        ];
    }

    private static ParsedGeoJsonFeature ReadFeature(
        JsonElement feature,
        int index,
        string fallbackName,
        int sourceSrid)
    {
        if (feature.ValueKind != JsonValueKind.Object)
        {
            return Failed(
                index,
                $"{fallbackName} {index + 1}",
                "The feature must be a GeoJSON object.");
        }

        var name = ReadFeatureName(
            feature,
            $"{fallbackName} {index + 1}");

        if (!feature.TryGetProperty("geometry", out var geometry) ||
            geometry.ValueKind is JsonValueKind.Null or
                JsonValueKind.Undefined)
        {
            return Failed(
                index,
                name,
                "The feature does not contain a geometry.");
        }

        return ReadGeometryFeature(
            geometry,
            index,
            name,
            sourceSrid);
    }

    private static ParsedGeoJsonFeature ReadGeometryFeature(
        JsonElement geometry,
        int index,
        string name,
        int sourceSrid)
    {
        try
        {
            var geometryType =
                ReadRequiredString(geometry, "type");

            if (!geometry.TryGetProperty(
                "coordinates",
                out var coordinates))
            {
                return Failed(
                    index,
                    name,
                    "The geometry does not contain coordinates.");
            }

            return geometryType switch
            {
                "Polygon" => ReadPolygon(
                    coordinates,
                    index,
                    name,
                    sourceSrid),
                "MultiPolygon" => ReadMultiPolygon(
                    coordinates,
                    index,
                    name,
                    sourceSrid),
                _ => Failed(
                    index,
                    name,
                    $"Geometry type '{geometryType}' is not supported. " +
                    "Use Polygon or MultiPolygon.")
            };
        }
        catch (SiteImportException exception)
        {
            return Failed(index, name, exception.Message);
        }
        catch (JsonException exception)
        {
            return Failed(
                index,
                name,
                $"The geometry coordinates are malformed: {exception.Message}");
        }
    }

    private static ParsedGeoJsonFeature ReadPolygon(
        JsonElement coordinates,
        int index,
        string name,
        int sourceSrid)
    {
        var rings =
            coordinates.Deserialize<double[][][]>()
            ?? throw Invalid(
                "content",
                "invalid_coordinates",
                "The Polygon coordinates cannot be null.");

        var warnings = new List<string>();
        var polygons = BuildPolygons(
            TransformIfRequired(rings, sourceSrid),
            warnings);

        return new ParsedGeoJsonFeature(
            index,
            name,
            polygons,
            null,
            warnings);
    }

    private static ParsedGeoJsonFeature ReadMultiPolygon(
        JsonElement coordinates,
        int index,
        string name,
        int sourceSrid)
    {
        var polygonCoordinates =
            coordinates.Deserialize<double[][][][]>()
            ?? throw Invalid(
                "content",
                "invalid_coordinates",
                "The MultiPolygon coordinates cannot be null.");
        var warnings = new List<string>();
        var polygons = polygonCoordinates
            .SelectMany(rings => BuildPolygons(
                TransformIfRequired(rings, sourceSrid),
                warnings))
            .ToArray();

        if (polygons.Length == 0)
        {
            return Failed(
                index,
                name,
                "The MultiPolygon does not contain any polygons.");
        }

        return new ParsedGeoJsonFeature(
            index,
            name,
            polygons,
            null,
            warnings);
    }

    private static IReadOnlyList<Polygon> BuildPolygons(
        double[][][] rings,
        List<string> warnings)
    {
        try
        {
            var polygons =
                GeoJsonPolygonMapper.ToImportPolygons(
                    new GeoJsonPolygonRequest(
                        "Polygon",
                        rings,
                        GeoJsonPolygonMapper.RequiredSrid),
                    out var repaired);

            if (repaired)
            {
                warnings.Add(
                    "An invalid polygon was repaired during import.");
            }

            return polygons;
        }
        catch (PolygonValidationException exception)
        {
            throw Invalid(
                "content",
                exception.Problem,
                exception.Message);
        }
    }

    private static double[][][] TransformIfRequired(
        double[][][] rings,
        int sourceSrid)
    {
        if (sourceSrid != 3857)
        {
            return rings;
        }

        return rings
            .Select(ring => ring
                .Select(position =>
                    TransformWebMercatorPosition(position))
                .ToArray())
            .ToArray();
    }

    private static double[] TransformWebMercatorPosition(
        double[] position)
    {
        if (position.Length != 2 ||
            !double.IsFinite(position[0]) ||
            !double.IsFinite(position[1]))
        {
            return position;
        }

        var longitude =
            position[0] / EarthRadiusMetres * 180 / Math.PI;
        var latitude =
            (2 * Math.Atan(
                Math.Exp(position[1] / EarthRadiusMetres)) -
             Math.PI / 2) *
            180 / Math.PI;

        return [longitude, latitude];
    }

    private static int DetectSrid(JsonElement root)
    {
        if (!root.TryGetProperty("crs", out var crs) ||
            crs.ValueKind is JsonValueKind.Null or
                JsonValueKind.Undefined)
        {
            return GeoJsonPolygonMapper.RequiredSrid;
        }

        string? name = null;

        if (crs.ValueKind == JsonValueKind.Object &&
            crs.TryGetProperty("properties", out var properties) &&
            properties.ValueKind == JsonValueKind.Object &&
            properties.TryGetProperty("name", out var nameElement) &&
            nameElement.ValueKind == JsonValueKind.String)
        {
            name = nameElement.GetString();
        }

        var normalized = name?.ToUpperInvariant() ?? string.Empty;

        if (normalized.Contains("3857", StringComparison.Ordinal) ||
            normalized.Contains("900913", StringComparison.Ordinal))
        {
            return 3857;
        }

        if (normalized.Contains("4326", StringComparison.Ordinal) ||
            normalized.Contains("CRS84", StringComparison.Ordinal))
        {
            return GeoJsonPolygonMapper.RequiredSrid;
        }

        throw Invalid(
            "content",
            "unsupported_crs",
            string.IsNullOrWhiteSpace(name)
                ? "The GeoJSON coordinate system could not be detected."
                : $"Coordinate system '{name}' is not supported. " +
                  "Use EPSG:4326 or EPSG:3857.");
    }

    private static string ReadFeatureName(
        JsonElement feature,
        string fallback)
    {
        if (!feature.TryGetProperty(
            "properties",
            out var properties) ||
            properties.ValueKind != JsonValueKind.Object)
        {
            return fallback;
        }

        foreach (var propertyName in new[] { "name", "title", "id" })
        {
            if (properties.TryGetProperty(
                    propertyName,
                    out var value) &&
                value.ValueKind == JsonValueKind.String &&
                !string.IsNullOrWhiteSpace(value.GetString()))
            {
                return value.GetString()!.Trim();
            }
        }

        return fallback;
    }

    private static string ReadRequiredString(
        JsonElement element,
        string propertyName)
    {
        if (element.ValueKind != JsonValueKind.Object ||
            !element.TryGetProperty(propertyName, out var value) ||
            value.ValueKind != JsonValueKind.String ||
            string.IsNullOrWhiteSpace(value.GetString()))
        {
            throw Invalid(
                "content",
                "invalid_geojson",
                $"GeoJSON '{propertyName}' must be a string.");
        }

        return value.GetString()!;
    }

    private static ParsedGeoJsonFeature Failed(
        int index,
        string name,
        string error)
    {
        return new ParsedGeoJsonFeature(
            index,
            name,
            [],
            error,
            []);
    }

    private static SiteImportException Invalid(
        string field,
        string problem,
        string message)
    {
        return new SiteImportException(
            field,
            problem,
            message);
    }
}

File: frontend/apps/site/src/app/features/sites/site-import/geojson-import.parser.ts
import { GeoJsonPolygon, GeoJsonPosition } from '@frontend/models';

const WEB_MERCATOR_RADIUS = 6_378_137;

interface GeoJsonRecord {
  [key: string]: unknown;
}

export interface GeoJsonImportPreviewFeature {
  index: number;
  name: string;
  polygons: GeoJsonPolygon[];
  issues: string[];
  warnings: string[];
}

export interface GeoJsonImportPreview {
  fileName: string;
  content: string;
  detectedCoordinateSystem: string;
  featureCount: number;
  invalidFeatureCount: number;
  features: GeoJsonImportPreviewFeature[];
}

export class GeoJsonPreviewError extends Error {}

export function parseGeoJsonImport(
  content: string,
  fileName: string,
): GeoJsonImportPreview {
  let root: unknown;

  try {
    root = JSON.parse(content);
  } catch {
    throw new GeoJsonPreviewError(
      'This file is not valid JSON. Check for missing commas or brackets.',
    );
  }

  if (!isRecord(root)) {
    throw new GeoJsonPreviewError('The GeoJSON root must be an object.');
  }

  const sourceSrid = detectSrid(root);
  const fallbackName =
    fileName.replace(/\.(geo)?json$/i, '').trim() || 'Imported site';
  const featureRecords = collectFeatureRecords(root);
  const features = featureRecords.map((feature, index) =>
    parseFeature(feature, index, fallbackName, sourceSrid),
  );

  return {
    fileName,
    content,
    detectedCoordinateSystem:
      sourceSrid === 3857
        ? 'Web Mercator (EPSG:3857) · will convert to WGS 84'
        : 'WGS 84 (EPSG:4326)',
    featureCount: features.length,
    invalidFeatureCount: features.filter(
      (feature) => feature.issues.length > 0,
    ).length,
    features,
  };
}

function collectFeatureRecords(root: GeoJsonRecord): GeoJsonRecord[] {
  const type = stringProperty(root, 'type');

  if (type === 'FeatureCollection') {
    if (!Array.isArray(root['features'])) {
      throw new GeoJsonPreviewError(
        'A FeatureCollection must contain a features array.',
      );
    }

    return root['features'].map((feature) => {
      if (!isRecord(feature)) {
        return { type: 'InvalidFeature' };
      }

      return feature;
    });
  }

  return [root];
}

function parseFeature(
  record: GeoJsonRecord,
  index: number,
  fallbackName: string,
  sourceSrid: number,
): GeoJsonImportPreviewFeature {
  const isFeature = record['type'] === 'Feature';
  const geometry = isFeature ? record['geometry'] : record;
  const name = isFeature
    ? readFeatureName(record, `${fallbackName} ${index + 1}`)
    : featureDefaultName(fallbackName, index);
  const issues: string[] = [];
  const warnings: string[] = [];
  const polygons: GeoJsonPolygon[] = [];

  if (!isRecord(geometry)) {
    issues.push('The feature does not contain a geometry.');
    return { index, name, polygons, issues, warnings };
  }

  const geometryType = geometry['type'];
  const coordinates = geometry['coordinates'];

  if (geometryType !== 'Polygon' && geometryType !== 'MultiPolygon') {
    issues.push(
      `Geometry type '${String(geometryType ?? 'unknown')}' is not supported. Use Polygon or MultiPolygon.`,
    );
    return { index, name, polygons, issues, warnings };
  }

  const polygonCoordinates =
    geometryType === 'Polygon' ? [coordinates] : coordinates;

  if (!Array.isArray(polygonCoordinates) || polygonCoordinates.length === 0) {
    issues.push('The geometry does not contain polygon coordinates.');
    return { index, name, polygons, issues, warnings };
  }

  for (const candidate of polygonCoordinates) {
    const parsed = parsePolygonCoordinates(candidate, sourceSrid);

    if (parsed.issue) {
      issues.push(parsed.issue);
    } else if (parsed.polygon) {
      polygons.push(parsed.polygon);

      if (parsed.warning) {
        warnings.push(parsed.warning);
      }
    }
  }

  if (polygons.length === 0 && issues.length === 0) {
    issues.push('The geometry does not contain a valid polygon.');
  }

  return { index, name, polygons, issues, warnings };
}

function parsePolygonCoordinates(
  value: unknown,
  sourceSrid: number,
): {
  polygon?: GeoJsonPolygon;
  issue?: string;
  warning?: string;
} {
  if (!Array.isArray(value) || value.length === 0) {
    return { issue: 'A Polygon must contain at least one ring.' };
  }

  const rings: GeoJsonPosition[][] = [];

  for (const ringValue of value) {
    if (!Array.isArray(ringValue) || ringValue.length < 4) {
      return {
        issue: 'Every polygon ring must contain at least four positions.',
      };
    }

    const ring: GeoJsonPosition[] = [];

    for (const positionValue of ringValue) {
      const position = parsePosition(positionValue, sourceSrid);

      if (!position) {
        return {
          issue:
            'Every position must contain finite [longitude, latitude] numbers in range.',
        };
      }

      ring.push(position);
    }

    if (!samePosition(ring[0], ring[ring.length - 1])) {
      return {
        issue:
          'Every polygon ring must be closed; its first and last positions must match.',
      };
    }

    rings.push(ring);
  }

  const hasSelfIntersection = rings.some((ring) => ringSelfIntersects(ring));

  return {
    polygon: {
      type: 'Polygon',
      coordinates: rings,
    },
    warning: hasSelfIntersection
      ? 'This polygon self-intersects. The importer will repair it when safe.'
      : undefined,
  };
}

function parsePosition(
  value: unknown,
  sourceSrid: number,
): GeoJsonPosition | null {
  if (
    !Array.isArray(value) ||
    value.length !== 2 ||
    typeof value[0] !== 'number' ||
    typeof value[1] !== 'number' ||
    !Number.isFinite(value[0]) ||
    !Number.isFinite(value[1])
  ) {
    return null;
  }

  const transformed =
    sourceSrid === 3857
      ? webMercatorToWgs84(value[0], value[1])
      : ([value[0], value[1]] as GeoJsonPosition);

  return transformed[0] >= -180 &&
    transformed[0] <= 180 &&
    transformed[1] >= -90 &&
    transformed[1] <= 90
    ? transformed
    : null;
}

function webMercatorToWgs84(
  x: number,
  y: number,
): GeoJsonPosition {
  const longitude = (x / WEB_MERCATOR_RADIUS) * (180 / Math.PI);
  const latitude =
    (2 * Math.atan(Math.exp(y / WEB_MERCATOR_RADIUS)) - Math.PI / 2) *
    (180 / Math.PI);

  return [longitude, latitude];
}

function detectSrid(root: GeoJsonRecord): number {
  if (root['crs'] === undefined || root['crs'] === null) {
    return 4326;
  }

  const crs = root['crs'];
  const properties = isRecord(crs) ? crs['properties'] : undefined;
  const name =
    isRecord(properties) && typeof properties['name'] === 'string'
      ? properties['name'].toUpperCase()
      : '';

  if (name.includes('3857') || name.includes('900913')) {
    return 3857;
  }

  if (name.includes('4326') || name.includes('CRS84')) {
    return 4326;
  }

  throw new GeoJsonPreviewError(
    name
      ? `Coordinate system '${name}' is not supported. Use EPSG:4326 or EPSG:3857.`
      : 'The GeoJSON coordinate system could not be detected.',
  );
}

function readFeatureName(
  feature: GeoJsonRecord,
  fallback: string,
): string {
  const properties = feature['properties'];

  if (!isRecord(properties)) {
    return fallback;
  }

  for (const key of ['name', 'title', 'id']) {
    const value = properties[key];

    if (typeof value === 'string' && value.trim()) {
      return value.trim();
    }
  }

  return fallback;
}

function featureDefaultName(
  fallbackName: string,
  index: number,
): string {
  return index === 0 ? fallbackName : `${fallbackName} ${index + 1}`;
}

function stringProperty(
  record: GeoJsonRecord,
  property: string,
): string {
  const value = record[property];

  if (typeof value !== 'string' || !value.trim()) {
    throw new GeoJsonPreviewError(
      `GeoJSON '${property}' must be a string.`,
    );
  }

  return value;
}

function ringSelfIntersects(ring: GeoJsonPosition[]): boolean {
  for (let first = 0; first < ring.length - 1; first++) {
    for (let second = first + 1; second < ring.length - 1; second++) {
      const adjacent =
        Math.abs(first - second) <= 1 ||
        (first === 0 && second === ring.length - 2);

      if (
        !adjacent &&
        segmentsIntersect(
          ring[first],
          ring[first + 1],
          ring[second],
          ring[second + 1],
        )
      ) {
        return true;
      }
    }
  }

  return false;
}

function segmentsIntersect(
  a: GeoJsonPosition,
  b: GeoJsonPosition,
  c: GeoJsonPosition,
  d: GeoJsonPosition,
): boolean {
  const orientation = (
    first: GeoJsonPosition,
    second: GeoJsonPosition,
    third: GeoJsonPosition,
  ) =>
    (second[1] - first[1]) * (third[0] - second[0]) -
    (second[0] - first[0]) * (third[1] - second[1]);

  const first = orientation(a, b, c);
  const second = orientation(a, b, d);
  const third = orientation(c, d, a);
  const fourth = orientation(c, d, b);

  return first * second < 0 && third * fourth < 0;
}

function samePosition(
  first: GeoJsonPosition,
  second: GeoJsonPosition,
): boolean {
  return first[0] === second[0] && first[1] === second[1];
}

function isRecord(value: unknown): value is GeoJsonRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
